core/src/main/java/org/apache/iceberg/FastAppend.java

```diff
@@ -192,6 +192,11 @@ protected void cleanUncommitted(Set<ManifestFile> committed) {
     }
   }
 
+  @Override
+  protected boolean cleanupAfterCommit() {
```

Comment:
While I think this is probably safe today, it seems like a change that is going to make the code more brittle because it isn't obvious when this should be overridden or how the attempts interact with the manifests that are written. For instance, if the behavior of `writeNewManifests` changes and creates intermediate work to clean up, how would someone know that this also needs to change?

I wonder if there's a better check here than number of attempts. What about checking whether the current set of manifests is valid and would be committed? `return !hasNewFiles && newManifests != null`?

Reply:
`cleanupAfterCommit` runs after commit, at that point `hasNewFiles=false and newManifests != null` so this check would not actually skip any work. 

I took a look at this further and noticed we could probably skip the READ of the metadata/READ of all manifests in SnapshotProducer without the need for classes to be cleanup-aware. See the new approach I took.

Reply:
The reason why `cleanUncommitted` must exist in `FastAppend` is because `cleanUncommitted` runs after both successful commit *and* unsuccessful commit. For successful commit, no work is needed. For unsuccessful commits that have exhausted retries, every file is deleted.

Reply:
Was slightly wrong above, `rewrittenAppendManifests` do need to be cleaned up. 

For example:
`appendManifest` is called, commit fails, we should cleanup the rewrittenAppendManifests.

Reply:
If the commit fails, when is `appendManifests` cleaned up? I thought that `writeNewManifests` would replace the append manifests, but if no commit succeeds then what cleans up?

Reply:
`appendManifests` shouldn't be cleaned up because they are not written by `FastAppend`. 

see existing comment in `cleanUncommitted` method
```
// clean up only rewrittenAppendManifests as they are always owned by the table
// don't clean up appendManifests as they are added to the manifest list and are not compacted
```

Reply:
Ah, sorry. I was mistaken. Thanks for clarifying!